transformer_engine/common/comm_gemm_overlap/comm_gemm_overlap.cpp

```diff
@@ -0,0 +1,953 @@
+  char *workspace_ptr = reinterpret_cast<char *>(workspace.dptr());
+  size_t workspace_size_chunk = workspace.numel() / _stream_compute.size();
+
+  // Catch up the main stream
+  NVTE_CHECK_CUDA(cudaEventRecord(_start_compute, stream_main));
+  NVTE_CHECK_CUDA(cudaStreamWaitEvent(_stream_send, _start_compute, 0));
+  NVTE_CHECK_CUDA(cudaStreamWaitEvent(_stream_recv, _start_compute, 0));
+  for (size_t i = 0; i < _stream_compute.size(); i++) {
+    NVTE_CHECK_CUDA(cudaStreamWaitEvent(_stream_compute[i], _start_compute, 0));
+  }
+
+  // GEMM and send/recv chunks
+  for (int i = 0; i < _tp_size; i++) {
+    // GEMM chunk
+    int input_b_chunk_id = (_tp_id + i + 1) % _tp_size;
+    char *input_b_chunk_ptr = input_b_ptr + (input_b_chunk_id * input_b_chunk_bytes);
+    auto input_b_chunk = TensorWrapper(reinterpret_cast<void *>(input_b_chunk_ptr), {n_chunk, k},
+                                       B.dtype(), nullptr, nullptr, B.scale_inv());
+    auto output_chunk =
+        TensorWrapper(ubufs[i].dptr(), ubufs[i].shape(), D.dtype(), D.amax(), D.scale(), nullptr);
+    int workspace_offset = (i % _stream_compute.size()) * workspace_size_chunk;
+    auto workspace_chunk = TensorWrapper(reinterpret_cast<void *>(workspace_ptr + workspace_offset),
+                                         {workspace_size_chunk}, workspace.dtype());
+
+    cudaStream_t gemm_stream =
+        (i == _tp_size - 1) ? stream_main : _stream_compute[i % _stream_compute.size()];
+    nvte_cublas_gemm(A.data(), input_b_chunk.data(), output_chunk.data(), bias.data(),
+                     pre_gelu_out.data(), A_trans, B_trans, grad, workspace_chunk.data(),
+                     accumulate, use_split_accumulator, _math_sms, gemm_stream);
+
+    if (i > 0) {
+      // P2P communication chunk
+      int send_offset = comm_bytes * (i - 1);
+      int recv_offset = comm_bytes * (i - 1 + _tp_size);
+      int send_rank = (_tp_id + i) % _tp_size + _rank_round_tp;
+      int recv_rank = (_tp_size + _tp_id - i) % _tp_size + _rank_round_tp;
+      NVTE_CHECK_CUDA(
+          cudaEventRecord(_start_comm, _stream_compute[(i - 1) % _stream_compute.size()]));
+      NVTE_CHECK_CUDA(cudaStreamWaitEvent(_stream_send, _start_comm, 0));
+      NVTE_CHECK_CUDA(cudaStreamWaitEvent(_stream_recv, _start_comm, 0));
+      ubuf::userbuffers_send(_ub_reg, send_offset, _ub_reg, recv_offset, comm_bytes, _ub_comm,
+                             send_rank, _stream_send);
+      ubuf::userbuffers_recv(_ub_reg, send_offset, _ub_reg, recv_offset, comm_bytes, _ub_comm,
+                             recv_rank, _stream_recv);
+    }
+  }
+
+  for (size_t i = 0; i < _stream_compute.size(); i++) {
+    NVTE_CHECK_CUDA(cudaEventRecord(_stop_compute, _stream_compute[i]));
+    NVTE_CHECK_CUDA(cudaStreamWaitEvent(stream_main, _stop_compute, 0));
+  }
+
+  NVTE_CHECK_CUDA(cudaEventRecord(_stop_send, _stream_send));
+  NVTE_CHECK_CUDA(cudaStreamWaitEvent(stream_main, _stop_send, 0));
+
```

Comment:
It seems [these 2 line (NVIDIA/TransformerEngine)](https://github.com/NVIDIA/TransformerEngine/blob/ee541e83b7088db1580907450a4f88491c8ccaf8/transformer_engine/pytorch/csrc/comm_gemm_overlap.h#L1236-L1237) is missing.
```cpp
NVTE_CHECK_CUDA(cudaEventRecord(_stop_recv, (cudaStream_t)_stream_recv));
NVTE_CHECK_CUDA(cudaStreamWaitEvent((cudaStream_t)stream_main, _stop_recv, 0));
```

The reduce part probably need to wait for all `ubuf::userbuffers_recv` to finish receiving data. Then, it can continue to reduce.
Not sure if I miss something. Does this make sense?